### ISSUE TITLE
dereferencing type members in uprobe fails with "X does not contain a field named Y"

### ISSUE THREAD

**Issue by laf0rge** [iovisor/bpftrace#2212]
Using bpftrace master of today (38c099dff9100bafeaaa7cee865f4dfda58134ac), I'm unable to dereference field members of structs:

My target shared library is built with -gdwarf -fno-omit-frame-pointer and bpftrace shows it understands the function arguments and their types:

```
# ./bpftrace -lv 'uprobe:/usr/local/lib/libosmocore.so.18.0.0:osmo_timer_schedule'
uprobe:/usr/local/lib/libosmocore.so.18.0.0:osmo_timer_schedule
    struct osmo_timer_list* timer
    int seconds
    int microseconds

```
However, when trying something like this:

```
uprobe:/usr/local/lib/libosmocore.so.18.0.0:osmo_timer_schedule /comm == "osmo-bsc"/ {
        $ms = arg1*1000 + arg2/1000;
        if ($ms >= 64 && $ms < 127) {
                printf("%p\n", args->timer->cb);
        }
}

```
I'm getting:

```
/root/timer-cb.bt:4:18-33: ERROR: Struct/union of type 'struct osmo_timer_list' does not contain a field named 'cb'
        printf("%p\n", args->timer->cb);
                       ~~~~~~~~~~~~~~~

```

However, there clearly is a 'cb' member, and it's also present in DWARF (output from objdum --dwarf onthe shared library):

```
<1><27a>: Abbrev Number: 11 (DW_TAG_structure_type)
    <27b>   DW_AT_name        : (indirect string, offset: 0x1f1): osmo_timer_list
    <27f>   DW_AT_byte_size   : 80
    <280>   DW_AT_alignment   : 8
    <281>   DW_AT_decl_file   : 6
    <282>   DW_AT_decl_line   : 52
    <283>   DW_AT_decl_column : 8
    <284>   DW_AT_sibling     : <0x2da>
 <2><288>: Abbrev Number: 12 (DW_TAG_member)
    <289>   DW_AT_name        : (indirect string, offset: 0x223): node
    <28d>   DW_AT_decl_file   : 6
    <28e>   DW_AT_decl_line   : 53
    <28f>   DW_AT_decl_column : 17
    <290>   DW_AT_type        : <0x21e>
    <294>   DW_AT_alignment   : 8
    <295>   DW_AT_data_member_location: 0
 <2><296>: Abbrev Number: 7 (DW_TAG_member)
    <297>   DW_AT_name        : (indirect string, offset: 0x354d): list
    <29b>   DW_AT_decl_file   : 6
    <29c>   DW_AT_decl_line   : 54
    <29d>   DW_AT_decl_column : 20
    <29e>   DW_AT_type        : <0x1dd>
    <2a2>   DW_AT_data_member_location: 24
 <2><2a3>: Abbrev Number: 7 (DW_TAG_member)
    <2a4>   DW_AT_name        : (indirect string, offset: 0x64ed): timeout
    <2a8>   DW_AT_decl_file   : 6
    <2a9>   DW_AT_decl_line   : 55
    <2aa>   DW_AT_decl_column : 17
    <2ab>   DW_AT_type        : <0x1a2>
    <2af>   DW_AT_data_member_location: 40
 <2><2b0>: Abbrev Number: 13 (DW_TAG_member)
    <2b1>   DW_AT_name        : (indirect string, offset: 0x1ea): active
    <2b5>   DW_AT_decl_file   : 6
    <2b6>   DW_AT_decl_line   : 56
    <2b7>   DW_AT_decl_column : 15
    <2b8>   DW_AT_type        : <0x16c>
    <2bc>   DW_AT_byte_size   : 4
    <2bd>   DW_AT_bit_size    : 1
    <2be>   DW_AT_bit_offset  : 31
    <2bf>   DW_AT_data_member_location: 56
 <2><2c0>: Abbrev Number: 14 (DW_TAG_member)
    <2c1>   DW_AT_name        : cb
    <2c4>   DW_AT_decl_file   : 6
    <2c5>   DW_AT_decl_line   : 58
    <2c6>   DW_AT_decl_column : 9
    <2c7>   DW_AT_type        : <0x2ea>
    <2cb>   DW_AT_data_member_location: 64
 <2><2cc>: Abbrev Number: 7 (DW_TAG_member)
    <2cd>   DW_AT_name        : (indirect string, offset: 0x31ab): data
    <2d1>   DW_AT_decl_file   : 6
    <2d2>   DW_AT_decl_line   : 59
    <2d3>   DW_AT_decl_column : 8
    <2d4>   DW_AT_type        : <0x1a0>
    <2d8>   DW_AT_data_member_location: 72
```

The C struct definition looks like this:
```
/*! A structure representing a single instance of a timer */
struct osmo_timer_list {
        struct rb_node node;      /*!< rb-tree node header */
        struct llist_head list;   /*!< internal list header */
        struct timeval timeout;   /*!< expiration time */
        unsigned int active  : 1; /*!< is it active? */

        void (*cb)(void*);        /*!< call-back called at timeout */
        void *data;               /*!< user data for callback */
};
```

What am I doing wrong here? IMHO all the information is there, but bpftrace falsely claims that a struct member named 'cb' doesn't exist.

**Comment by viktormalik** [comment]
The problem is the `active` field. The DWARF parser does not yet support bitfields and it falls back to the Clang parser which needs the full definition of the type. Bitfield support is on my todo list, but I didn't get to it, yet.

Hopefully I'll fix that soon, in the meantime, you can bypass the error by embedding the struct def in your script.

**Comment by viktormalik** [comment]
Hi, I finally managed to fix this issue in #2505. Mind giving it a try? Thanks!
